### DISCUSSION TITLE
Machine learning API

### DISCUSSION THREAD

**Discussion by libkush** [immich-app/immich#8497] [Q&A] [answered]
I am very impressed with the facial recognition capabilities of Immich. I am building a personal app (strictly for personal use only), so my questions are -

a) Is it legally allowed to copy paste the entire machine learning service from Immich under the AGPL-3.0 license? If so, am I allowed to just use it internally with friends and family without having to open source it?
b) If it is allowed, what exactly does an API request and a response look like to the machine learning microservice (it would be really helpful if you can provide an example), and what are the other requirements I need to take care of before trying to integrate in my crappy project. I have read some details about this service [here](https://github.com/immich-app/immich/discussions/8347) and I think this is precisely what I need, .i.e, send HTTP request with an image, get face embeddings and bounding boxes as response.

PS. I know this sounds like I'm trying to steal code and even seeking help in doing it, but I promise this is not for any competitor or commercial app but a really important project I need for personal reasons. However, if you deem this as inappropriate, I will avoid copying your implementation.

**Comment by bo0tzz** [comment]
Under AGPL, if you make any changes at all to the code you are required to make that changed code available under AGPL as well. Of course if you don't make changes, or then just use the container image that we publish, you don't need to open source anything. 

The ML API isn't documented anywhere, you're probably best off looking at https://github.com/immich-app/immich/blob/main/server/src/repositories/machine-learning.repository.ts for how we use it.

**Comment by mertalev** [comment, marked as answer]
a) Since you're creating a derivative work (code that depends on this service), you need to make the complete code available to any users of the service. This includes instructions on where to get it and making it available online if the usage is online (which sounds like the case). The license of the overall work also has to be a flavor of GPL (it technically doesn't need to be AGPL v3, but the network clause will apply to the full work anyway). However, it doesn't need to be accessible by non-users. IANAL; this is just my understanding of the license.

b) A simple example for usage is [locustfile.py](https://github.com/immich-app/immich/blob/main/machine-learning/locustfile.py). It creates a dummy image, but in normal usage the image would be loaded from a file.

**Comment by LackesLab** [comment]
@zettachad The original source of the models is the insightface repository. 
https://github.com/deepinsight/insightface

You can check the models and code there. They have some nice code examples to start with
